util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,238 @@
+        "collection": collection,
+        "database": database,
+    }
+    data = json.dumps(data).encode("utf-8")
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get resources: {e}")
+
+    return json.loads(response.read().decode("utf-8"))["documents"]
+
+
+def save_resources_to_file(resources: List, output: pathlib.Path):
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def progress_hook(t):
+    last_b = [0]
+
+    def update_to(b=1, bsize=1, tsize=None):
+        if tsize not in (None, -1):
+            t.total = tsize
+        displayed = t.update((b - last_b[0]) * bsize)
+        last_b[0] = b
+        return displayed
+
+    return update_to
+
+
+def download_resources(output: pathlib.Path):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    with open(output.joinpath("resources.json")) as f:
+        resources = json.load(f)
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filename = pathlib.Path(
+                *pathlib.Path(parse.urlsplit(url).path).parts[3:]
+            )
+            filepath = pathlib.Path(path).joinpath(filename)
+            filepath.parent.mkdir(parents=True, exist_ok=True)
+
+            if (
+                not filepath.exists()
+                or hashlib.md5(filepath.read_bytes()).hexdigest()
+                != resource["md5sum"]
+            ):
+                with tqdm(
+                    unit="B",
+                    unit_scale=True,
+                    unit_divisor=1024,
+                    miniters=1,
+                    desc=f"Downloading {url}",
+                ) as t:
+                    request.urlretrieve(url, filepath, progress_hook(t))
+            resource["url"] = filepath.absolute().as_uri()
+
+    with open(output.joinpath("resources.json"), "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+if __name__ == "__main__" or __name__ == "__m5_main__":
+    parser = argparse.ArgumentParser(
+        description="Get resources from the database"
+    )
+
+    parser.add_argument(
+        "--config-file-url",
+        type=str,
+        default="https://raw.githubusercontent.com/gem5/gem5/stable/"
```

Comment:
I still think that this should be a path to a local file and not an URL.
We want this script to download the resources currently used in the user setup.

Reply:
Agreed. 

So the script doesn't get bogged down with relative paths ( we should be able to move the script around and not have some hardcoded paths break), perhaps we can use a hardlink here so this file exists both in "util/offline_db" and "src/python/". Then we just have this default set to the directory the script is in plus "gem5_default_confg.py" (I don't actually know if git supports hardlinks ... you should check that before proceeding).